### ISSUE TITLE
`Local Definition` vs `Let`

### ISSUE THREAD

**Issue by nmvdw** [UniMath/UniMath#1456]
This discussion came up twice:
- https://github.com/UniMath/UniMath/pull/1432#discussion_r776996433
- https://github.com/UniMath/UniMath/pull/1453#discussion_r804739418

As said by @rmatthes , it would be good to have a global discussion about their respective (dis)advantages. It might also be good to adjust the style guide  accordingly.

**Comment by rmatthes** [comment]
I can point to a potential disadvantages of `Let` by example. The axiom [enriched_functor_comp_ax](https://github.com/UniMath/UniMath/blob/master/UniMath/CategoryTheory/Enriched/Enriched.v#L206) is defined with several `Let`, and if one prints it out, four let-bindings for identifiers that are not on display for the formula are still listed: `(I := monoidal_cat_unit Mon_V) (α := monoidal_cat_associator Mon_V) (l_unitor := monoidal_cat_left_unitor Mon_V) (r_unitor := monoidal_cat_right_unitor Mon_V)`.

There is anyway a word of warning in order: if one uses wildcards (underscore) in the definition of statements, the system is not supposed to synthesize the arguments by using the concepts that are abbreviated by  `Let`. As an example, the lemma [right_unitor_inv_of_tensor](https://github.com/UniMath/UniMath/blob/master/UniMath/CategoryTheory/Monoidal/MonoidalCategories.v#L474) - the last one in its file. It states `(id _ #⊗ nat_z_iso_to_trans_inv r_unitor _) = nat_z_iso_to_trans_inv r_unitor (X ⊗ Y)  · α ((_, _), _)`, where the very last wildcard stands for the unit of the monoidal category. The print output is then `# (monoidal_cat_tensor Mon_V)
         (id X #, nat_z_iso_to_trans_inv (monoidal_cat_right_unitor Mon_V) Y) =
       nat_z_iso_to_trans_inv (monoidal_cat_right_unitor Mon_V)
         (monoidal_cat_tensor Mon_V (X, Y))
       · monoidal_cat_associator Mon_V ((X, Y), pr1 (pr2 (pr2 Mon_V)))` which does not expand the let-binding for the unit but just gives the sequence of projections. Of course, one should just write I for _ in this situation. Otherwise, the behaviour of `Let` is as expected: in the proven formula, the let-bindings are all unfolded and thus behave like aliases.

**Comment by benediktahrens** [comment]
@rmatthes : thanks a lot for giving these examples; they illustrate some issues with `Let` very clearly.

I currently think (but am very happy to have my opinion changed) that neither `Let` nor `Local Definition` are particularly well-suited, in particular regarding the case we are discussing: defining projections out of a mathematical structure.
Instead, it seems much better to me to give a global definition with a potentially cumbersome name, and to define notations, in a separate submodule, that can - and needs to - be opened separately from importing the definitions themselves.

I believe that @nmvdw has used this in some places - could you point us to an example?

**Comment by nmvdw** [comment]
I use `Let` for abbreviations. I don't use it with arguments and I don't use it for stuff that is supposed to be opaque (I prefer `Local Lemma` for that). The abbreviations are needed for stuff in a particular section, but not for stuff that's needed elsewhere. See for example: https://github.com/UniMath/UniMath/blob/master/UniMath/Bicategories/Morphisms/Properties/ClosedUnderPullback.v#L103

I prefer it over `Local Definition`, because it doesn't add anything to the name space. I think `Let` is useful when adding something that is only needed at one particular place.

**Comment by rmatthes** [comment]
Concerning notations: am I right that ProofGeneral (and Company-Coq) are still not so good in uncovering the notations that are used in a given expression in the source code? E.g., in case of our symbol · for composition of morphisms that appears in more than 5kloc?

**Comment by rmatthes** [comment]
To make my last remark more understandable: If I type C-c C-a C-n in PG when the cursor is over the composition symbol, Emacs wants to "Locate notation" with default search string consisting only of that symbol. When hitting RET, this yields a syntax error. This does not happen for `->`,  `-->`, `==>`, `λ`, `∏`, `∑`, `#`, `⋆`. And it does not happen for the composition symbol when doing the operation at the [line](https://github.com/UniMath/UniMath/blob/master/UniMath/CategoryTheory/Core/Functors.v#L1080) where this notation is introduced. The problem also occurs with `⟹`, `•`, `◃`, `▹`. Maybe it is a bug of PG or even of my installation (with Company-Coq).
How would you resolve the notations for functor categories in the form `[_,_]` or `[_,_,_]` or for hom-sets in the form `_⟦_, _⟧`?

**Comment by DanGrayson** [comment]
As I recall, sometimes proof-general fails to put quotation marks around the symbol in the command to Coq, resulting in a syntax error.

**Comment by rmatthes** [comment]
Okay, is this a long-standing problem? As I wrote, this can depend on the context even for a given symbol.
For my final question above, I already have as solution typing as search strings ` "[ _ , _ ]"`, "`[ _ , _ , _ ]"` and `"_ ⟦ _ , _ ⟧"`, always with enough spacing. What I did not know is that C-c C-a C-b with these search strings in PG even displays where the concepts for which the notations stand are defined. Is there a way to point to where the notations are defined?

**Comment by DanGrayson** [comment]
I don't see one.  If I do C-c C-a C-n (coq-LocateNotation) with "[" I get just this display:
```coq
Notation
"[ C , D ]" := functor_category C D : cat (default interpretation)
"[ C , D , hs ]" := functor_precategory C D hs : cat (default interpretation)
```

No locations.

But I like using `M-x tags-search` -- use a regular expression such as `^Notation.*\[.*,.*\]` and you'll quickly find it.

By the way, it's a good idea to "reserve" all new notations in Foundations/Init.v -- that way we don't get two files using different precedences for the same notation, which will be an error.

**Comment by rmatthes** [comment]
The locations came from C-c C-a C-b with those search strings for notation.

**Comment by DanGrayson** [comment]
Oh.  That seems to give the definition, without giving the location of the "Notation" command.

**Comment by rmatthes** [comment]
@DanGrayson : Yes, the C-c C-a C-b only shows the location of the concept that is denoted by the notation, not the location of the notation. The Emacs tags facility that is prepared for use with UniMath and described in the [installation instructions](https://github.com/UniMath/UniMath/blob/master/INSTALL.md#tags-files) is indeed of real help here. I still wonder if I overlook a genuine Coq functionality for this purpose. If this is not so, I would be very glad to see your hint (including your regex - even if it does not distinguish between definiens and definiendum) in the UniMath documentation. INSTALL.md is not the place where one would expect it, maybe USAGE.md, or even better in a new file ADVANCEDUSAGE.md (if maybe USAGE.md is meant to show first steps to beginners).

**Comment by DanGrayson** [comment]
It turns out the INSTALL.md already had a section on TAGS file usage, so I've moved the tags-search discussion from USAGE.md to there.
